chapter/methodology.md

```diff
@@ -0,0 +1,321 @@
+
+Once fully built a `Building` instance is post-processed yielding a `Completed` linearization.
+While being defined similar to its origin, the structure is optimized for positional access, affecting the order of the `LinearizationItem`s and requiring an auxiliary mapping for efficient access to items by their `id`.
+Moreover, types of items in the `Completed` linearization will be resolved.
+
+Type definitions of the `Linearization` as well as its type-states `Building` and `Completed` are listed in [@lst:nickel-definition-lineatization;@lst:nls-definition-building-type;@lst:nls-definition-completed-type].
+Note that only the former is defined as part of the Nickel libraries, the latter are specific implementations for NLS. 
+
+
+```{.rust #lst:nickel-definition-lineatization caption="Definition of Linearization structure"}
+pub trait LinearizationState {}
+
+pub struct Linearization<S: LinearizationState> {
+    pub state: S,
+}
+```
+
+```{.rust #lst:nls-definition-building-type caption="Type Definition of Building state"}
+pub struct Building {
+    pub linearization: Vec<LinearizationItem<Unresolved>>,
+    pub scope: HashMap<Vec<ScopeId>, Vec<ID>>,
+}
+impl LinearizationState for Building {}
+```
+
+```{.rust #lst:nls-definition-completed-type caption="Type Definition of Completed state" }
+pub struct Completed {
+    pub linearization: Vec<LinearizationItem<Resolved>>,
+    scope: HashMap<Vec<ScopeId>, Vec<usize>>,
+    id_to_index: HashMap<ID, usize>,
+}
+impl LinearizationState for Completed {}
+```
+
+
+
+### Transfer from AST
+
+#### Retyping
+
+### Post-Processing
+
+### Resolving Elements
+
+#### Resolving by position
+
+As part of the post-processing step discussed in [@sec:post-processing], the `LinearizationItem`s in the `Completed` linearization are reorderd by their occurence of the corresponding AST node in the source file.
+To find items in this list three preconditions have to hold:
+
+
+1. Each element has a corresponding span in the source
+2. Items of different files appear ordered by `FileId`
+3. Spans may overlap never intersect.
+   $$\text{Item}^2_\text{start} \geq \text{Item}^1_\text{start} \land \text{Item}^2_\text{end} \leq \text{Item}^1_\text{end}$$
+4. Items referring to the spans starting at the same position have to occur in the same order before and after the post-processing. 
+   Concretely, this ensures that the tree-induced hierarchy is maintained, more precise elements follow broader ones
+
+
+This first two properties are an implication of the preceding processes.
+All elements are derived from AST nodes, which are parsed from files retaining their position.
+Nodes that are generated by the runtime before being passed to the language server are either ignored or annotated with synthetic positions that are known to be in the bounds of the file and meet the second requirement.
+For all other nodes the second requirement is automatically fulfilled by the grammar of the Nickel language.
```

Comment:
I'm sorry, I think I lost the original conversations and it seems you have re-ordered items since then, because now 2. is

> 2. Items of different files appear ordered by `FileId`

What was the point that is automatically ensured by the Nickel grammar? Also, your example seems to be about the dynamic of the language, which I suspect is a different matter. Grammar-wise, the JavaScript example and the corresponding Nickel translation would parse in the same way.

Reply:
I think the comment was on whether the Nickel grammar implies that scopes are only nested but not overlapping

Reply:
Then, scopes are not grammar related, they appear later. Put differently, the parser doesn't care about scopes, all it do is to have occurrences of `x` in the example above, which could very well be translated directly to Nickel. The fact that different occurrences of `x` refer or not to the same variable semantically is part of the language operation semantics. I don't think saying "Nickel's grammar" or LALR grammars have this property or not really makes sense, to begin with.